### PR TITLE
KAFKA-14132: Replace PowerMock and EasyMock with Mockito in connect tests

### PR DESCRIPTION
**Pull request by clolov**
Batch 2 of the tests detailed in https://issues.apache.org/jira/browse/KAFKA-14132 which use PowerMock/EasyMock and need to be moved to Mockito.

### CODE REVIEW

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java [hunk 1]
```diff
@@ -71,122 +74,74 @@ public void setup() {
                 Long.toString(DEFAULT_OFFSET_COMMIT_INTERVAL_MS));
         WorkerConfig config = new StandaloneConfig(workerProps);
         committer = new SourceTaskOffsetCommitter(config, executor, committers);
-        Whitebox.setInternalState(SourceTaskOffsetCommitter.class, "log", mockLog);
```
Comment (clolov) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java:74]: I did not manage to find such a functionality in Mockito, so I decided to get rid of it.

**Comment by clolov** [comment]
Hey @C0urante! Tagging you for visibility as this is a Connect-related test.

**Comment by C0urante** [comment]
Hi @clolov - thanks for the PR! The title mentions streams tests; should that be updated?

I plan to finish the first round of review sometime within the next week.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java (hunk 1 above)
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java:74]: I'm not a fan of using Whitebox to mess around with private fields, but we should find a way to preserve the coverage guarantees that we get from these tests w/r/t logging. Left a comment below with a suggestion on how we can do that.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java [hunk 2]
```diff
@@ -71,122 +74,74 @@ public void setup() {
...
     public void testClose() throws Exception {
         long timeoutMs = 1000;
 
         // Normal termination, where termination times out.
-        executor.shutdown();
-        PowerMock.expectLastCall();
-
-        EasyMock.expect(executor.awaitTermination(eq(timeoutMs), eq(TimeUnit.MILLISECONDS)))
-                .andReturn(false);
-        mockLog.error(EasyMock.anyString());
```
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java:107]: We want to verify here that at least one `ERROR`-level log message was emitted by the `SourceTaskOffsetCommitter` instance during this part of the test.

We can use the `LogCaptureAppender` class to do that exact thing, without installing a mock logger for the class:
```java
        try (LogCaptureAppender logCaptureAppender = LogCaptureAppender.createAndRegister(SourceTaskOffsetCommitter.class)) {
            committer.close(timeoutMs);
            assertTrue(logCaptureAppender.getEvents().stream().anyMatch(e -> e.getLevel().equals("ERROR")));
        }
```

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java [hunk 3]
```diff
@@ -71,122 +74,74 @@ public void setup() {
...
-        PowerMock.expectLastCall();
-        PowerMock.replayAll();
+        when(executor.awaitTermination(timeoutMs, TimeUnit.MILLISECONDS)).thenReturn(false);
 
         committer.close(timeoutMs);
```
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java]: Can we add a `verify(executor, times(1)).shutdown();` after this line, to ensure that the executor has actually been shut down by this point?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java [hunk 4]
```diff
@@ -71,122 +74,74 @@ public void setup() {
...
-        PowerMock.verifyAll();
-        PowerMock.resetAll();
-
         // Try to remove a cancelled task
-        EasyMock.expect(taskFuture.cancel(eq(false))).andReturn(false);
-        EasyMock.expect(taskFuture.isDone()).andReturn(false);
-        EasyMock.expect(taskFuture.get()).andThrow(new CancellationException());
-        EasyMock.expect(taskId.connector()).andReturn("MyConnector");
-        EasyMock.expect(taskId.task()).andReturn(1);
-        mockLog.trace(EasyMock.anyString(), EasyMock.<Object>anyObject());
```
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java:162]: We can also use the `LogCaptureAppender` class here. Will probably have to invoke `LogCaptureAppender.setClassLoggerToTrace(SourceTaskOffsetCommitter.class);`

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java [hunk 5]
```diff
@@ -71,122 +74,74 @@ public void setup() {
...
+        when(executor.awaitTermination(timeoutMs, TimeUnit.MILLISECONDS)).thenThrow(new InterruptedException());
 
         committer.close(timeoutMs);
 
-        PowerMock.verifyAll();
+        verify(executor, times(2)).shutdown();
     }
 
     @Test
     public void testRemove() throws Exception {
```
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java]: We're losing some coverage guarantees by restructuring the tests like this. Right now we know for each sub-case (removing a task successfully, removing a cancelled one, and being interrupted while removing one) that we call `cancel` and `get` on the future. With the changes here, we do get guarantees that those methods are invoked at least once at some point in the `testRemove` method (thanks to strict stubbing), but we don't get finer-grained guarantees than that. And ensuring that those methods are invoked exactly when we expect is fairly valuable, so it'd be nice if we could keep that coverage.

I think we can break this test case out into four separate cases:
1. Removing a non-existent task
2. Removing a task successfully
3. Removing a task that gets cancelled
4. Removing a task and getting interrupted while doing so

And we can use a helper method to reduce duplication for the last three:
```java
private void testRemove() {
    // Try to remove an existing task
    when(taskFuture.cancel(false)).thenReturn(false);
    when(taskFuture.isDone()).thenReturn(false);
    when(taskId.connector()).thenReturn("MyConnector");
    when(taskId.task()).thenReturn(1);

    committers.put(taskId, taskFuture);
    committer.remove(taskId);
}
```

Which would make each test case fairly brief. For example, case 2 might look like this:

```java
@Test
public void testRemoveSuccess() throws Exception {
    when(taskFuture.get()).thenReturn(null);
    testRemove();
    assertEquals(Collections.emptyMap(), committers);
}
```

This way, after each test case completes, strict stubbing will ensure that the expected methods were invoked on the future at least once. We could go further and explicitly verify that they were invoked exactly once, but up to you if you'd like to add that coverage or not.

**Review by C0urante — COMMENTED** [review]
Thanks @clolov! This was shorter than I remember, was able to finish a round today.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java (hunk 2 above)
Comment (clolov) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java:107]: This is a very fair and clean suggestion. I have added it to the subsequent commits.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java (hunk 3 above)
Comment (clolov) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java]: I have opted for splitting this test case into two. Verify should now be present in both.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java (hunk 5 above)
Comment (clolov) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java]: I have split this into 4 tests as you suggested. I have a slight preference for only abstracting the setup of the mocks, but if you would prefer to have as little duplication as possible I will change it.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java (hunk 4 above)
Comment (clolov) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java:162]: Amended in the subsequent commit!

**Comment by clolov** [comment]
Hey @C0urante! Thanks a lot for the review. Hopefully I have addressed the first batch of comments :)

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java [hunk 6]
```diff
@@ -71,112 +75,83 @@ public void setup() {
...
-        assertTrue(taskWrapper.hasCaptured());
         assertNotNull(taskWrapper.getValue());
         assertEquals(singletonMap(taskId, commitFuture), committers);
-
-        PowerMock.verifyAll();
     }
 
     @Test
-    public void testClose() throws Exception {
+    public void testCloseWithinTimeout() throws Exception {
```
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java]: Nit: this name seems to go against the case that we're covering here (where `close` fails to complete within the expected timeout). Maybe rename to `testCloseTimeout`?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java [hunk 7]
```diff
@@ -71,112 +75,83 @@ public void setup() {
...
+        }
 
-        committer.close(timeoutMs);
+        verify(executor).shutdown();
+    }
 
-        PowerMock.verifyAll();
-        PowerMock.resetAll();
+    @Test
+    public void testCloseOutsideOfTimeout() throws InterruptedException {
```
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java]: Nit: "outside of timeout" is a little unclear; maybe just `testCloseInterrupted`?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java [hunk 8]
```diff
@@ -185,8 +160,13 @@ public void testRemove() throws Exception {
         } catch (ConnectException e) {
             //ignore
         }
+    }
 
-        PowerMock.verifyAll();
+    private void removeSetup() {
```
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java]: Nit: "remove setup" kind of sounds like "remove" is a verb and we're removing something. Maybe `expectRemove`, which would also line up nicely with other utility functions in other tests?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java [hunk 9]
```diff
@@ -185,8 +160,13 @@ public void testRemove() throws Exception {
         } catch (ConnectException e) {
             //ignore
         }
```
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java]: We can replace this with `assertThrows`:
```java
        committers.put(taskId, taskFuture);
        assertThrows(ConnectException.class, () -> committer.remove(taskId));
```

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java (hunk 5 above)
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java]: Looks great!

**Review by C0urante — COMMENTED** [review]
Thanks @clolov! This is looking great; just a few small comments left and then it should be good to go.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java (hunk 6 above)
Comment (clolov) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java]: Fair point, changed in the next commit.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java (hunk 7 above)
Comment (clolov) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java]: I agree, I have changed it in the subsequent commit.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java (hunk 9 above)
Comment (clolov) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java]: Nice catch, changed in the subsequent commit.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java (hunk 8 above)
Comment (clolov) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitterTest.java]: Yup, makes a lot of sense. Changed in the next commit :)

**Review by C0urante — APPROVED** [review]
Great job, LGTM

**Comment by C0urante** [comment]
Changes in last commit were trivially verifiable, and the CI run for the commit before that did not contain any relevant test failures. Merging...

**Comment by clolov** [comment]
Thank you very much for the quick turnaround @C0urante!

**Pull request merged**